### DISCUSSION TITLE
Orbit Restructuring

### DISCUSSION THREAD

**Discussion by merks** [eclipse-orbit/orbit#49] [General]
Orbit has moved to its own GitHub organization:

- https://github.com/eclipse-orbit

In this organization, the new [orbit-simrel](https://github.com/eclipse-orbit/orbit-simrel#readme) repository is intended to replace (as much as possible) the use of [EBR](https://github.com/eclipse-orbit/ebr#readme) for the [obit](https://github.com/eclipse-orbit/orbit#readme) repository's recipes.

# maven-osgi

Every Orbit recipe that corresponds to a Maven dependency that is available from Maven central as an OSGi artifact, will be removed from the orbit repository. Projects can instead directly specify such dependencies in their target platform.  The following folder of the orbit-simrel repository shows how that's properly done, including the necessary Tycho infrastructure:

- https://github.com/eclipse-orbit/orbit-simrel/tree/main/maven-osgi

As a convenience for [SimRel](https://ci.eclipse.org/simrel/) participants, the Orbit project provides the following p2 repository with PGP-signed artifacts that is effectively the "augmented union" of all the dependencies of all SimRel participants:

- **https://download.eclipse.org/tools/orbit/simrel/maven-osgi/**

The dependencies here are automatically updated to the latest minor version as such versions become available.

In addition, reports are generated to help SimRel participants keep their dependencies up-to-date:

- https://github.com/eclipse-orbit/orbit-simrel/tree/main/report

Note that often the direct-from-Maven OSGi-bundle will have a different bundle symbolic name than the legacy Orbit bundled version.

# maven-bnd

Each remaining Orbit recipe will gradually be replaced (as much as possible) by a dependency specified in a PDE target platform using a "maven" location along with BND instructions.  Here's an example:

https://github.com/eclipse-orbit/orbit-simrel/blob/9a03cd1aba89351daccf1d69a7e3d6f4f268bf84/maven-bnd/tp/MavenBND.target#L6-L83

The Orbit project provides the following p2 repository with jar-signed bundled artifacts:

- **https://download.eclipse.org/tools/orbit/simrel/maven-bnd/**

The dependencies here too are automatically updated to the latest minor version as such versions become available.

# maven-sign

Some bundles contain native libraries.  Native macos libraries, `*.jnilib`, in particular must generally be signed in order to be able to notarize products containing such bundles.

The Orbit project provides the following p2 repository with jar-signed bundled artifacts containing signed native libraries:

- **https://download.eclipse.org/tools/orbit/simrel/maven-sign/**

The dependencies here too are automatically updated to the latest minor version as such versions become available.

# orbit-aggregation

To provide "one-stop-shopping", the Orbit project provides a p2 repository that is an aggregation of the following:

- Traditional EBR-wrapped OSGi bundles.
- Maven direct OSGi bundles.
- Maven BND-wrapped OSGi bundles.
- Maven bundles with signed native libraries.
- Legacy OSGi bundles required by the above.

I.e., the following aggregated repository is provided:

- **https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/**

# Repository Structure

This is the general structure of all these three new p2 repositories:

- https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/nightly/latest
- https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/milestone/latest  **- after first milestone**
- https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/release/latest  **- after first release**
- https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/2023-09  **- latest nightly, milestone or release**

# Legacy Links

The existing "legacy composites are planned to be updated as follows:

- https://download.eclipse.org/tools/orbit/downloads/latest-I
  - https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/nightly/latest
- https://download.eclipse.org/tools/orbit/downloads/latest-S
  - https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/2023-09
     - In preparation for 2023-09 M2 - approximately July 25:
- https://download.eclipse.org/tools/orbit/downloads/latest-R
  - https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/2023-09
    - Final build for 2023-09 (R-build) - approximately Sep 5:

This cycle will repeat.

_Note that old releases (2023-06 and earlier) URLs will not be modified and will continue to point at the same location they were already pointing at._

# SimRel Recommendations

- It is highly recommended to use package imports to specify the dependencies of bundles, particularly dependencies on 3rd party libraries.  
- It is highly recommended that you stop including (or even importing) 3rd party libraries in your features, particularly in the features that you contribute to the SimRel aggregation.
  - In order to continue to include such libraries in your update site:
    - Include them directly in the category.xml.
    - Define a 3rd party libraries feature, include them there, and include that in your category.xml.

Following the above recommendations will allow your contribution to be aggregated without duplication and will allow your contributed bundles to resolve their requirements to newer updated versions (within the range constraints you specify) as those versions become available.  Collectively, this will help us to simplify and streamline Orbit's management of 3rd party libraries, In addition, it will make it easier to be responsive to any CVEs that might be discovered in 3rd party libraries without you needing to lift a finger!

# Finding Things

Many of the bundles that comes directly from Maven as OSGi artifacts have a different bundle symbolic name than before.  The generated page for each simple update site includes useful links and details to help you find what you need.  You can click the underlined swizzle to expand all the details to make it easier to search for things like  package names.  The bundles that are available directly from Maven, and can be specified directly in your target platform using a `maven` location, have a link to Maven Central which gives you the coordinates that you can use:

![image](https://github.com/eclipse-orbit/orbit/assets/208716/a74f4b17-a00c-4464-9536-1b719a9d8af8)

The Repository Explorer is helpful too, especially in "Expert Mode" as underlined on the view toolbar, along with filtering as underlined:

![image](https://github.com/eclipse-orbit/orbit/assets/208716/0870c10c-ba2b-456d-ba01-398074c540cb)

You can even search for packages and double click on one to open the details dialog where you can see which bundles provide that package.

![image](https://github.com/eclipse-orbit/orbit/assets/208716/b2ff9971-ebb4-4528-9173-a3bc288514bd)

**Comment by cdietrich** [comment]
Many thanks for your efforts @merks @jonahgraham

**Reply by ruspl-afed** [reply]
A lot of issues regarding third-party use can now be resolved, thank you so much @merks @jonahgraham !

**Comment by jonahgraham** [comment]
I added _Note that old releases (2023-06 and earlier) URLs will not be modified and will continue to point at the same location they were already pointing at._ to the Legacy Links section to make it clear that we aren't modifying old Orbit URLs.
